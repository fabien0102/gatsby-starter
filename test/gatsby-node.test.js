/* eslint-disable no-undef, max-nested-callbacks */
jest.mock('path');
const {createPages, onCreateNode} = require('../gatsby-node');

describe('gatsby-node', () => {
  const boundActionCreators = {};

  describe('createPages', () => {
    let graphql;

    beforeEach(() => {
      boundActionCreators.createPage = jest.fn();
      graphql = jest.fn();
    });

    it('should create blog posts', () => {
      graphql.mockReturnValueOnce(
        Promise.resolve(
          {
            data: {
              posts: {
                edges: [
                  {
                    node: {
                      fields: {
                        slug: '/blog/2017-04-18--welcoming/'
                      }
                    }
                  },
                  {
                    node: {
                      fields: {
                        slug: '/blog/2017-05-02--article-2/'
                      }
                    }
                  },
                  {
                    node: {
                      fields: {
                        slug: '/blog/2017-05-02--article-1/'
                      }
                    }
                  },
                  {
                    node: {
                      fields: {
                        slug: '/docs/dont-take-me/'
                      }
                    }
                  }
                ]
              }
            }
          }
        )
      );

      return createPages({graphql, boundActionCreators})
        .then(() => {
          expect(boundActionCreators.createPage.mock.calls).toMatchSnapshot();
        });
    });

    it('should throw an error on graphql error', () => {
      graphql.mockReturnValueOnce(
        Promise.resolve({errors: 'something wrong!'})
      );

      expect(createPages({graphql, boundActionCreators}))
        .toThrow();
    });

    it('should create tags pages', () => {
      graphql.mockReturnValueOnce(
        Promise.resolve(
          {
            data: {
              posts: {
                edges: [
                  {
                    node: {
                      fields: {
                        slug: '/blog/2017-04-18--welcoming/'
                      },
                      frontmatter: {
                        tags: [
                          'starter',
                          'gatsby'
                        ]
                      }
                    }
                  },
                  {
                    node: {
                      fields: {
                        slug: '/blog/2017-05-02--article-2/'
                      },
                      frontmatter: {
                        tags: [
                          'test'
                        ]
                      }
                    }
                  },
                  {
                    node: {
                      fields: {
                        slug: '/blog/2017-05-02--article-1/'
                      },
                      frontmatter: {
                        tags: [
                          'test'
                        ]
                      }
                    }
                  },
                  {
                    node: {
                      fields: {
                        slug: '/blog/2017-05-02--article-0/'
                      },
                      frontmatter: {
                        tags: null
                      }
                    }
                  }
                ]
              }
            }
          }
        )
      );

      return createPages({graphql, boundActionCreators})
        .then(() => {
          expect(boundActionCreators.createPage.mock.calls).toMatchSnapshot();
        });
    });

    describe('pagination', () => {
      const generateData = n => {
        const edges = [...Array(n)].map((val, i) => ({
          node: {
            fields: {
              slug: `/blog/2017-04-18--article-${i + 1}/`
            }
          }
        }));

        return {data: {posts: {edges}}};
      };

      it('should create 1 page with 5 posts', () => {
        graphql.mockReturnValueOnce(Promise.resolve(generateData(5)));
        return createPages({graphql, boundActionCreators})
          .then(() => {
            const pages = boundActionCreators.createPage.mock.calls
              .filter(d => d[0].path && d[0].path.startsWith('/blog/page/'));
            expect(pages).toMatchSnapshot();
          });
      });

      it('should create 2 pages with 15 posts', () => {
        graphql.mockReturnValueOnce(Promise.resolve(generateData(15)));
        return createPages({graphql, boundActionCreators})
          .then(() => {
            const pages = boundActionCreators.createPage.mock.calls
              .filter(d => d[0].path && d[0].path.startsWith('/blog/page/'));
            expect(pages).toMatchSnapshot();
          });
      });

      it('should create 3 pages with 30 posts', () => {
        graphql.mockReturnValueOnce(Promise.resolve(generateData(30)));
        return createPages({graphql, boundActionCreators})
          .then(() => {
            const pages = boundActionCreators.createPage.mock.calls
              .filter(d => d[0].path && d[0].path.startsWith('/blog/page/'));
            expect(pages).toMatchSnapshot();
          });
      });
    });
  });

  describe('onCreateNode', () => {
    let getNode;

    beforeEach(() => {
      boundActionCreators.createNodeField = jest.fn();
      getNode = jest.fn();
    });

    it('should create slugs for MarkdownRemark file', () => {
      getNode.mockReturnValue(
        {
          relativePath: 'blog/2017-04-18--welcoming/index.md'
        }
      );

      const node = {
        internal: {
          type: 'MarkdownRemark'
        },
        parent: 'parent',
        frontmatter: {
          title: 'Welcoming',
          createdDate: '2017-04-18',
          updatedDate: '2017-05-06',
          author: 'Fabien BERNARD',
          tags: ['starter', 'gatsby'],
          image: 'pexels-photo-253092.jpeg',
          draft: false
        }
      };
      onCreateNode({node, boundActionCreators, getNode});

      expect(boundActionCreators.createNodeField.mock.calls).toMatchSnapshot();
    });

    it('should use frontmatter slug for MarkdownRemark file when defined', () => {
      getNode.mockReturnValue(
        {
          relativePath: 'blog/2017-04-18--welcoming/index.md'
        }
      );

      const node = {
        internal: {
          type: 'MarkdownRemark'
        },
        parent: 'parent',
        frontmatter: {
          title: 'Welcoming',
          createdDate: '2017-04-18',
          updatedDate: '2017-05-06',
          author: 'Fabien BERNARD',
          tags: ['starter', 'gatsby'],
          image: 'pexels-photo-253092.jpeg',
          draft: false,
          slug: '2017-04-18/welcoming'
        }
      };
      onCreateNode({node, boundActionCreators, getNode});

      expect(boundActionCreators.createNodeField.mock.calls).toMatchSnapshot();
    });

    it('should do nothing on unknown type', () => {
      getNode.mockReturnValue(
        {
          relativePath: 'blog/2017-04-18--welcoming/index.md'
        }
      );
      const node = {
        internal: {
          type: 'unknown'
        },
        parent: 'parent'
      };
      onCreateNode({node, boundActionCreators, getNode});

      expect(boundActionCreators.createNodeField.mock.calls.length).toBe(0);
    });
  });
});
