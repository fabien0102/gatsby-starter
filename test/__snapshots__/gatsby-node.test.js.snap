// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatsby-node createPages pagination should create 1 page with 5 posts 1`] = `
Array [
  Array [
    Object {
      "component": "base-path/src/templates/blog-page.tsx",
      "context": Object {
        "skip": 0,
      },
      "path": "/blog/page/1/",
    },
  ],
]
`;

exports[`gatsby-node createPages pagination should create 2 pages with 15 posts 1`] = `
Array [
  Array [
    Object {
      "component": "base-path/src/templates/blog-page.tsx",
      "context": Object {
        "skip": 0,
      },
      "path": "/blog/page/1/",
    },
  ],
  Array [
    Object {
      "component": "base-path/src/templates/blog-page.tsx",
      "context": Object {
        "skip": 10,
      },
      "path": "/blog/page/2/",
    },
  ],
]
`;

exports[`gatsby-node createPages pagination should create 3 pages with 30 posts 1`] = `
Array [
  Array [
    Object {
      "component": "base-path/src/templates/blog-page.tsx",
      "context": Object {
        "skip": 0,
      },
      "path": "/blog/page/1/",
    },
  ],
  Array [
    Object {
      "component": "base-path/src/templates/blog-page.tsx",
      "context": Object {
        "skip": 10,
      },
      "path": "/blog/page/2/",
    },
  ],
  Array [
    Object {
      "component": "base-path/src/templates/blog-page.tsx",
      "context": Object {
        "skip": 20,
      },
      "path": "/blog/page/3/",
    },
  ],
]
`;

exports[`gatsby-node createPages should create blog posts 1`] = `
Array [
  Array [
    Object {
      "component": "base-path/src/templates/blog-post.tsx",
      "context": Object {
        "slug": "/blog/2017-04-18--welcoming/",
      },
      "path": "/blog/2017-04-18--welcoming/",
    },
  ],
  Array [
    Object {
      "component": "base-path/src/templates/blog-post.tsx",
      "context": Object {
        "slug": "/blog/2017-05-02--article-2/",
      },
      "path": "/blog/2017-05-02--article-2/",
    },
  ],
  Array [
    Object {
      "component": "base-path/src/templates/blog-post.tsx",
      "context": Object {
        "slug": "/blog/2017-05-02--article-1/",
      },
      "path": "/blog/2017-05-02--article-1/",
    },
  ],
  Array [
    Object {
      "component": "base-path/src/templates/blog-page.tsx",
      "context": Object {
        "skip": 0,
      },
      "path": "/blog/page/1/",
    },
  ],
]
`;

exports[`gatsby-node createPages should create tags pages 1`] = `
Array [
  Array [
    Object {
      "component": "base-path/src/templates/blog-post.tsx",
      "context": Object {
        "slug": "/blog/2017-04-18--welcoming/",
      },
      "path": "/blog/2017-04-18--welcoming/",
    },
  ],
  Array [
    Object {
      "component": "base-path/src/templates/blog-post.tsx",
      "context": Object {
        "slug": "/blog/2017-05-02--article-2/",
      },
      "path": "/blog/2017-05-02--article-2/",
    },
  ],
  Array [
    Object {
      "component": "base-path/src/templates/blog-post.tsx",
      "context": Object {
        "slug": "/blog/2017-05-02--article-1/",
      },
      "path": "/blog/2017-05-02--article-1/",
    },
  ],
  Array [
    Object {
      "component": "base-path/src/templates/blog-post.tsx",
      "context": Object {
        "slug": "/blog/2017-05-02--article-0/",
      },
      "path": "/blog/2017-05-02--article-0/",
    },
  ],
  Array [
    Object {
      "component": "base-path/src/templates/tags-page.tsx",
      "context": Object {
        "tag": "starter",
      },
      "path": "/blog/tags/starter/",
    },
  ],
  Array [
    Object {
      "component": "base-path/src/templates/tags-page.tsx",
      "context": Object {
        "tag": "gatsby",
      },
      "path": "/blog/tags/gatsby/",
    },
  ],
  Array [
    Object {
      "component": "base-path/src/templates/tags-page.tsx",
      "context": Object {
        "tag": "test",
      },
      "path": "/blog/tags/test/",
    },
  ],
  Array [
    Object {
      "component": "base-path/src/templates/blog-page.tsx",
      "context": Object {
        "skip": 0,
      },
      "path": "/blog/page/1/",
    },
  ],
]
`;

exports[`gatsby-node onCreateNode should create slugs for MarkdownRemark file 1`] = `
Array [
  Array [
    Object {
      "name": "slug",
      "node": Object {
        "frontmatter": Object {
          "author": "Fabien BERNARD",
          "createdDate": "2017-04-18",
          "draft": false,
          "image": "pexels-photo-253092.jpeg",
          "tags": Array [
            "starter",
            "gatsby",
          ],
          "title": "Welcoming",
          "updatedDate": "2017-05-06",
        },
        "internal": Object {
          "type": "MarkdownRemark",
        },
        "parent": "parent",
      },
      "value": "/blog/2017-04-18--welcoming/",
    },
  ],
]
`;

exports[`gatsby-node onCreateNode should use frontmatter slug for MarkdownRemark file when defined 1`] = `
Array [
  Array [
    Object {
      "name": "slug",
      "node": Object {
        "frontmatter": Object {
          "author": "Fabien BERNARD",
          "createdDate": "2017-04-18",
          "draft": false,
          "image": "pexels-photo-253092.jpeg",
          "slug": "2017-04-18/welcoming",
          "tags": Array [
            "starter",
            "gatsby",
          ],
          "title": "Welcoming",
          "updatedDate": "2017-05-06",
        },
        "internal": Object {
          "type": "MarkdownRemark",
        },
        "parent": "parent",
      },
      "value": "/blog/2017-04-18/welcoming/",
    },
  ],
]
`;
